Add `--format` for `docker node ls`

yongtang opened the pull request:
**- What I did**
This fix tries to address the comment https://github.com/docker/docker/pull/30376#discussion_r97465334 where it was not possible to specify `--format` for `docker node ls`. The `--format` flag is a quite useful flag that could be used in many places such as completion.

**- How I did it**

This fix implements `--format` for `docker node ls` and add `nodesFormat` in config.json so that it is possible to specify the output when `docker node ls` is invoked.

Related documentations have been updated.

**- How to verify it**

A set of unit tests have been added.

**- Description for the changelog**

Add `--format` for `docker node ls` and add `nodesFormat` in config.json so that it is possible to specify the output when `docker node ls` is invoked.

**- A picture of a cute animal (not mandatory but encouraged)**

![enhanced-1301-1438290165-1](https://cloud.githubusercontent.com/assets/6932348/22267546/053b5822-e239-11e6-93f4-09439cab30be.jpg)


This fix is related to #30376.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add `--format` for `docker node ls`

This fix tries to address the comment https://github.com/docker/docker/pull/30376#discussion_r97465334
where it was not possible to specify `--format` for `docker node ls`. The `--format` flag
is a quite useful flag that could be used in many places such as completion.

This fix implements `--format` for `docker node ls` and add `nodesFormat` in config.json
so that it is possible to specify the output when `docker node ls` is invoked.

Related documentations have been updated.

A set of unit tests have been added.

This fix is related to #30376.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
/cc @albers @AkihiroSuda @vdemeester @thaJeztah

AkihiroSuda commented on docs/reference/commandline/node_ls.md:
````diff
@@ -120,6 +121,34 @@ ID                           HOSTNAME        STATUS  AVAILABILITY  MANAGER STATU
+`ID` and `Hostname` entries separated by a colon for all nodes:
+
+```bash
+$ docker plugin ls --format "{{.ID}}: {{.Hostname}}"
````
plugin???

yongtang replied:
@AkihiroSuda My bad. The PR has been updated. Thanks.

AkihiroSuda commented on cli/command/formatter/node_test.go:
```diff
@@ -0,0 +1,194 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	for i, line := range strings.Split(strings.TrimSpace(out.String()), "\n") {
```
We could deduplicate similar code in `cli/command/formatter/*_test.go`. But it can be another PR in the future.

AkihiroSuda commented:
SGTM

albers commented:
Thanks @yongtang. Your're doing great work!

vdemeester commented on cli/internal/test/cli.go:
```diff
@@ -42,7 +44,12 @@ func (c *FakeCli) Out() *command.OutStream {
+
+// ConfigFile returns the config file the cli will use
+func (c *FakeCli) ConfigFile() *configfile.ConfigFile {
+	return config.NewConfigFile("configFilename")
```
I choose another path for that in https://github.com/docker/docker/compare/master...vdemeester:volume-unit-tests#diff-ecd4d24831d477dbaef6c122698aa6ecR69 👼

yongtang replied:
Thanks @vdemeester. The PR has been updated for `cli/internal/test/cli.go`.

vdemeester reviewed (commented):
Design LGTM 👼
I would have added a case in the `node_list_test.go` tests, but otherwise looks good !

yongtang commented:
Thanks @vdemeester for the review. The PR has been updated with additional unit tests added in `node/list_test.go`. Please take a look.

vdemeester reviewed (approved):
LGTM 🐯

AkihiroSuda commented on docs/reference/commandline/node_ls.md:
````diff
@@ -120,6 +121,35 @@ ID                           HOSTNAME        STATUS  AVAILABILITY  MANAGER STATU
+
+```bash
+$ docker node ls --format "{{.ID}}: {{.Hostname}}"
+e216jshn25ckzbvmwlnh5jr3g *: swarm-manager1
````
Users are unlikely to expect ` *` to be included in `{{.ID}}`.
I understand this is by design, but maybe it is worth documenting that?

vdemeester replied:
I also think the `*` should be it's own element 👼.
@thaJeztah wdyt ?

thaJeztah replied:
Looks like I missed your ping here. Yes, I agree, the `*` should be a separate element. If it's included in the `ID`, it makes formatting a lot less useful (e.g., if people are automating things and want to extract the `ID`)

yongtang commented:
Thanks @AkihiroSuda. The PR has been rebased and updated. Now an additional note has been added:
```udiff
 
+> **Note:**
+> If the `ID` field of the node is followed by a `*` (e.g., `e216jshn25ckzbvmwlnh5jr3g *`)
+> in the above example output, then this node is also the node of the current docker daemon.
+
 
 ## Filtering
 
@@ -120,6 +125,35 @@ ID                           HOSTNAME        STATUS  AVAILABILITY  MANAGER STATU
 e216jshn25ckzbvmwlnh5jr3g *  swarm-manager1  Ready   Active        Leader

```

Please take a look and let me know if there are any issues.

AkihiroSuda commented:
@thaJeztah PTAL about the `*` symbol?

@yongtang please rebase?

yongtang commented:
@AkihiroSuda Thanks. The PR has been rebased.

@AkihiroSuda @vdemeester @thaJeztah I think to add `*` as an individual column, we might be able to achieve that by having a boolean placeholder with empty header name and either `"*"` or `""` as the output (`true` -> `"*"` and `false` -> `""`). Though I would like to get some feedback before moving forward.

vdemeester commented:
@yongtang how does it look with an empty header ? (I would think it's ok to go that way 👼)

yongtang commented:
@vdemeester I just tried with empty/whitespace header `" "` locally. It does not look good however. The reason is that no matter what spacing is added to the header, the current template implementation will always use `"\t"` to separate the headers. Therefore the header row and the remaining row are mismatched. 

Another option is to stick with `"\t"` delim but that looks ugly as well:
```
ID                                      HOSTNAME            STATUS              AVAILABILITY        MANAGER STATUS
nodeID1             *                   foobar_baz          Foo                 Drain               Leader
nodeID2                                 foobar_bar          Bar                 Active              Reachable		
```

There is a separate PR #30733 tracking the header delim of the template issue https://github.com/docker/docker/pull/28213#issuecomment-273840405 though. That PR may help address the separate 1-char column display (like `" "` or `"*"`).

allencloud commented on cli/command/node/list.go:
```diff
@@ -55,61 +48,27 @@ func runList(dockerCli command.Cli, opts listOptions) error {
+		if len(dockerCli.ConfigFile().NodesFormat) > 0 && !opts.quiet {
+			format = dockerCli.ConfigFile().NodesFormat
+		} else {
+			format = formatter.TableFormatKey
```
Could we just use 
```
format = formatter.TableFormatKey
if len(dockerCli.ConfigFile().NodesFormat) > 0 && !opts.quiet {
    format = dockerCli.ConfigFile().NodesFormat
}
```
Since it seems that most of us do not like `else`? 😄  just a nit.

yongtang replied:
@allencloud Updated. Thanks.

yongtang commented:
Rebased to fix merge conflict.

yongtang committed:
Add hidden placeholder of `.Self` for `docker node ls --format`

This commit adds a hidden placeholder of `.Self` for
`docker node ls --format` so that if the node is the same
as the current docker daemon, then a `*` is outputed.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks all for the review. The PR has been updated. Now there is a one-char hidden column/placeholder of `.Self`. For this placeholder, if the node is the same as the docker daemon, then a `*` will be emitted, otherwise, a `" "` (one-char whitespace) is emitted. The header of this hidden placeholder is also `" "`.

Below is the example:
```
ubuntu@ubuntu:~/docker$ docker node ls
ID                            HOSTNAME            STATUS              AVAILABILITY        MANAGER STATUS
hgfg1nagl0pl10ci0wgy4api7 *   ubuntu              Ready               Active              Leader
ubuntu@ubuntu:~/docker$ docker node ls --format 'table {{.ID}}: {{.Self}}'
ID:  
hgfg1nagl0pl10ci0wgy4api7: *
ubuntu@ubuntu:~/docker$ 
```

Not sure this is the best way to handle the scenario though with the merge of #30733 it is at least possible now to have a one-char placeholder/column.

Please take a look and any suggestions are welcomed.

AkihiroSuda commented on cli/command/formatter/node.go:
```diff
@@ -0,0 +1,104 @@
+	defaultNodeTableFormat = "table {{.ID}} {{.Self}}\t{{.Hostname}}\t{{.Status}}\t{{.Availability}}\t{{.ManagerStatus}}"
+
+	nodeIDHeader        = "ID"
+	selfHeader          = " "
```
an empty string could be used here

yongtang replied:
Thanks @AkihiroSuda for the review. The PR has been updated. Please take a look.

yongtang commented:
Rebased to fix merge conflict.

tiborvass commented:
LGTM

tiborvass commented:
Ping @thaJeztah @mstanleyjones @johndmulhausen

vdemeester commented on docs/reference/commandline/node_ls.md:
````diff
@@ -45,6 +46,11 @@ ID                           HOSTNAME        STATUS  AVAILABILITY  MANAGER STATU
 e216jshn25ckzbvmwlnh5jr3g *  swarm-manager1  Ready   Active        Leader
 ```
+> **Note:**
+> In the above example output, there is a hidden colume of `.Self` that indicates if the
````
s/colume/column/

yongtang replied:
Thanks @vdemeester. The typo has been fixed.

thaJeztah commented:
I'm just thinking if `.Self` should be a Boolean to separate "presentation" from "data" (e.g. `.IsCurrent`); that way, a JSON format output would possibly make more sense, and users can pick their own "format". It _does_ make the template a bit more difficult to use though, so open to input. Template would probably look something like;

```
table {{.ID}}{{if .IsCurrent}} * {{end}}\t{{.Hostname}}\t{{.Status}}\t{{.Availability}}\t{{.ManagerStatus}}
```

Given, we can already to `{{if eq .Self "*"}} yay, I am "self"{{ end }}`, so happy to hear if we're ok with the current implementation :blush:

thaJeztah commented:
ping @yongtang, I was briefly discussing this, and perhaps it's best to use the `boolean` approach if possible; would you be able to have a look if that works?

yongtang commented on docs/reference/commandline/cli.md:
```diff
@@ -167,6 +167,11 @@ property is not set, the client falls back to the default table
 
+The property `nodesFormat` specifies the default format for `docker node ls` output.
+When the `--format` flag is not provided with the `docker node ls` command,
+Docker's client uses this property. If this property is not set, the client
```
Thanks @mstanleyjones. Done.

yongtang commented on docs/reference/commandline/cli.md:
```diff
@@ -167,6 +167,11 @@ property is not set, the client falls back to the default table
+The property `nodesFormat` specifies the default format for `docker node ls` output.
+When the `--format` flag is not provided with the `docker node ls` command,
+Docker's client uses this property. If this property is not set, the client
+falls back to the default table format. For a list of supported formatting
```
Done.

yongtang commented on docs/reference/commandline/node_ls.md:
````diff
@@ -45,6 +46,11 @@ ID                           HOSTNAME        STATUS  AVAILABILITY  MANAGER STATU
 38ciaotwjuritcdtn9npbnkuz    swarm-worker1   Ready   Active
 e216jshn25ckzbvmwlnh5jr3g *  swarm-manager1  Ready   Active        Leader
 ```
+> **Note:**
````
Done.

yongtang commented:
Thanks @thaJeztah. The PR has been updated. Now the default format is 
```
defaultNodeTableFormat = "table {{.ID}} {{if .Self}}*{{else}} {{ end }}\t{{.Hostname}}\t{{.Status}}\t{{.Availability}}\t{{.ManagerStatus}}"
```
which should achieve the same effect while at the same time allows `.Self` to be boolean. Please take a look.

mdlinville reviewed (approved):
Docs LGTM

vdemeester reviewed (approved):
LGTM 🦁

thaJeztah commented:
Heh, found a funny one. Not sure yet if it's a "bug" or just a bad template;

```
docker node ls --format 'table {{.ID}} {{if not .Self}}*{{else}} {{ end }}\t{{.Hostname}}'
ID *                          HOSTNAME
ylm27u3p0wo04sdluj2ikigtf     6be31c5d1969
```

`if not .Self` is also evaluated in the "header" row, causing an `*` to be rendered there

Don't think that's a show stopper, but wanted to leave it as a comment in case someone has thoughts on how to (and _if_) to address :smile:

thaJeztah commented:
This is really nice to have with the boolean 👍 thanks!

```bash
docker node ls --format '{{json .}}' | jq .
{
  "Availability": "Active",
  "Hostname": "6be31c5d1969",
  "ID": "ylm27u3p0wo04sdluj2ikigtf",
  "ManagerStatus": "Leader",
  "Self": true,
  "Status": "Ready"
}
```

thaJeztah reviewed (approved):
LGTM, thanks so much for doing this (and bearing with my nits) 👍

thaJeztah commented:
/cc @albers @sdurrheimer :smile:

vdemeester commented:
🎉
